src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DeviceAuthHelper.cs

```diff
@@ -77,11 +75,99 @@ public static string GetBypassChallengeResponse(Dictionary<string, string> respo
+                challengeData["Version"]);
+        }
+
+        public static X509Certificate2 FindCertificate(IDictionary<string, string> challengeData)
+        {
+            var store = new X509Store(StoreName.My, StoreLocation.CurrentUser);
+            try
+            {
+                store.Open(OpenFlags.ReadOnly);
+                var certCollection = store.Certificates;
+                if (challengeData.ContainsKey("CertAuthorities"))
+                {
+                    return FindCertificateByCertAuthorities(challengeData, certCollection);
+                }
+
+                X509Certificate2Collection signingCert = null;
+                signingCert = certCollection.Find(X509FindType.FindByThumbprint, challengeData["CertThumbprint"],
+                    false);
+                if (signingCert.Count == 0)
+                {
+                    throw new MsalException(MsalError.DeviceCertificateNotFound,
+                        string.Format(CultureInfo.CurrentCulture, MsalErrorMessage.DeviceCertificateNotFoundTemplate,
+                            "Cert thumbprint:" + challengeData["CertThumbprint"]));
+                }
+
+                return signingCert[0];
+            }
+            finally
+            {
+
+#if NETSTANDARD || WINDOWS_APP
+                store.Dispose();
+#else
+                store.Close();
+#endif
+            }
+        }
+
+        private static X509Certificate2 FindCertificateByCertAuthorities(IDictionary<string, string> challengeData, X509Certificate2Collection certCollection)
+        {
+            X509Certificate2Collection signingCert = null;
+            string[] certAuthorities = challengeData["CertAuthorities"].Split(new[] { ";" },
+                StringSplitOptions.None);
+            foreach (var certAuthority in certAuthorities)
+            {
+                //reverse the tokenized string and replace "," with " + "
+                string[] dNames = certAuthority.Split(new[] { "," }, StringSplitOptions.None);
+                string distinguishedIssuerName = dNames[dNames.Length - 1];
+                for (int i = dNames.Length - 2; i >= 0; i--)
+                {
+                    distinguishedIssuerName += " + " + dNames[i].Trim();
+                }
+
+                signingCert = certCollection.Find(X509FindType.FindByIssuerDistinguishedName,
+                    distinguishedIssuerName, false);
+                if (signingCert.Count > 0)
+                {
+                    break;
+                }
+            }
+
+            if (signingCert == null || signingCert.Count == 0)
+            {
+                throw new MsalException(MsalError.DeviceCertificateNotFound,
+                    string.Format(CultureInfo.CurrentCulture,
+                        MsalErrorMessage.DeviceCertificateNotFoundTemplate,
+                        "Cert Authorities:" + challengeData["CertAuthorities"]));
+            }
+
+            return signingCert[0];
+        }
     }
```

Comment:
was this code added or moved from another file? wondering how closely to review it. thanks.

Reply:
@jennyf19 Yes, this was just extracted from the platform specific AuthManager classes.

Reply:
ok cool. thanks. i won't review then..assume it's already been reviewed. thx

---
In reply to: [607430889](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2521#discussion_r607430889) [](ancestors = 607430889)